core/src/main/java/com/datastrato/gravitino/storage/relational/service/UserMetaService.java

```diff
@@ -0,0 +1,237 @@
+              return;
+            }
+            SessionUtils.doWithoutCommit(
+                UserRoleRelMapper.class,
+                mapper -> {
+                  if (overwritten) {
+                    mapper.batchInsertUserRoleRelOnDuplicateKeyUpdate(userRoleRelPOs);
+                  } else {
+                    mapper.batchInsertUserRoleRel(userRoleRelPOs);
+                  }
+                });
+          });
+    } catch (RuntimeException re) {
+      ExceptionUtils.checkSQLException(
+          re, Entity.EntityType.USER, userEntity.nameIdentifier().toString());
+      throw re;
+    }
+  }
+
+  public boolean deleteUser(NameIdentifier identifier) {
+    Preconditions.checkArgument(
+        identifier != null
+            && !identifier.namespace().isEmpty()
+            && identifier.namespace().levels().length == 3,
+        "The identifier should not be null and should have three level.");
+    Long metalakeId =
+        MetalakeMetaService.getInstance().getMetalakeIdByName(identifier.namespace().level(0));
+    Long userId = getUserIdByMetalakeIdAndName(metalakeId, identifier.name());
+
+    SessionUtils.doMultipleWithCommit(
+        () ->
+            SessionUtils.doWithoutCommit(
+                UserMetaMapper.class, mapper -> mapper.softDeleteUserMetaByUserId(userId)),
+        () ->
+            SessionUtils.doWithoutCommit(
+                UserRoleRelMapper.class, mapper -> mapper.softDeleteUserRoleRelByUserId(userId)));
+    return true;
+  }
+
+  public <E extends Entity & HasIdentifier> UserEntity updateUser(
+      NameIdentifier identifier, Function<E, E> updater) {
+    Preconditions.checkArgument(
+        identifier != null
+            && !identifier.namespace().isEmpty()
+            && identifier.namespace().levels().length == 3,
+        "The identifier should not be null and should have three level.");
+
+    Long metalakeId =
+        MetalakeMetaService.getInstance().getMetalakeIdByName(identifier.namespace().level(0));
+    UserPO oldUserPO = getUserPOByMetalakeIdAndName(metalakeId, identifier.name());
+    List<RolePO> rolePOs = RoleMetaService.getInstance().listRolesByUserId(oldUserPO.getUserId());
+    UserEntity oldUserEntity = POConverters.fromUserPO(oldUserPO, rolePOs, identifier.namespace());
+
+    UserEntity newEntity = (UserEntity) updater.apply((E) oldUserEntity);
+    Preconditions.checkArgument(
+        Objects.equals(oldUserEntity.id(), newEntity.id()),
+        "The updated user entity id: %s should be same with the user entity id before: %s",
+        newEntity.id(),
+        oldUserEntity.id());
+
+    Set<Long> oldRoleIds =
+        oldUserEntity.roleIds() == null
+            ? Sets.newHashSet()
+            : Sets.newHashSet(oldUserEntity.roleIds());
+    Set<Long> newRoleIds =
+        newEntity.roleIds() == null ? Sets.newHashSet() : Sets.newHashSet(newEntity.roleIds());
+
+    Set<Long> grantRoleIds = Sets.difference(newRoleIds, oldRoleIds);
```

Comment:
grantRoleIds -> insertRoleIds.
revokeRoleIds -> deleteRoleIds.
Grant / revoke is the concept of upper layer.